carnot/carnot_vote_aggregation.py

```diff
@@ -0,0 +1,421 @@
+        assert block.id() in self.safe_blocks
+        # This assertion will be moved outside as the approve_block will be called in two cases:
+        # 1st the fast path when len(votes) == self.overlay.super_majority_threshold(self.id) and the second
+        # When there is the first timeout t1 for the fast path and the protocol operates in the slower path
+        # in this case the node will prepare a QC from votes it has received.
+        # assert len(votes) == self.overlay.super_majority_threshold(self.id)
+        assert all(self.overlay.is_member_of_subtree(self.id, vote.voter) for vote in votes)
+        assert all(vote.block == block.id() for vote in votes)
+        assert self.highest_voted_view < block.view
+
+        # Create a QC based on committee membership
+        qc = self.build_qc(block.view, block, None)  # if self.overlay.is_member_of_root_committee(self.id) else None
+
+        # Create a new vote
+        vote = Vote(
+            block=block.id(),
+            voter=self.id,
+            view=block.view,
+            qc=qc
+        )
+
+        # Update the highest voted view
+        self.highest_voted_view = max(self.highest_voted_view, block.view)
+
+        # After block verification, votes are sent to committee members.
+        # When a QC is formed from 2/3rd of subtree votes, it's forwarded to the parent committee.
+        # If a Type 1 timeout occurs, a QC is built from available votes and QCs and sent to the parent.
+        # Subsequent votes are forwarded to the parent committee members.
+        if self.overlay.is_member_of_root_committee():
+            recipient = self.overlay.leader(block.view + 1)
+        else:
+            recipient = self.overlay.my_committee(self.id)
+
+        # Return a Send event to the appropriate recipient
+        return Send(to=recipient, payload=vote)
+
+    # NewView msgs are not needed anymore
+    def build_qc(self, view: View, block: Optional[Block], timeouts: Optional[Set[Timeout]]) -> Qc:
+        # unhappy path
+        if timeouts:
+            timeouts = list(timeouts)
+            return AggregateQc(
+                qcs=[msg.high_qc.view for msg in timeouts],
+                highest_qc=max(timeouts, key=lambda x: x.high_qc.view).high_qc,
+                view=timeouts[0].view
+            )
+        # happy path
+        return StandardQc(
+            view=view,
+            block=block.id()
+        )
+
+    # A node initially forwards a vote or qc from its subtree to its parent committee. There can be two instances this
+    # can happen: 1: If a node forms a QC qc from votes and QCs it receives from its subtree such that the total number of votes in the qc is at two-third of votes from the subtree, then
+    # it forwards this QC to the parent committee members or a subset of parent committee members.
+    # 2: After sending the qc any additional votes are forwarded to the parent committee members or a subset of parent committee members.
```

Comment:
this can be rewritten as a separate function with separate conditions from case 1